### ISSUE TITLE
Type conversion for n-D scalable vectors (n > 1)

### ISSUE THREAD

**Issue by banach-space** [iree-org/iree#14294]
[Implements #13334 ]

I was trying to clarify a restriction on type conversions for scalable vectors, specifically for types like: 
* `vector<[16]x[16]xi8>`, or 
* `vector<[16]x[16]xi8>`,

Types such as `vector<[16]x[16]xi8>`  would normally lead to arrays of scalable vectors, but that's not supported by LLVM ATM (https://llvm.org/docs/LangRef.html#global-variables). So, IIUC, we should never attempt converting `vector<[16]x[16]xi8>` to LLVM . However, the following operation makes a lot of sense in the context of Arm's SME:
```
vector.transfer_write %cst, %arg0[%c0, %c0] {in_bounds = [true, true]} : vector<[16]x[16]xi8>, memref<?x?xi8>
```
In fact, the lowering for `vector.transfer_write` to SME was added in: https://reviews.llvm.org/D152508.

**Q:** Before `transfer_write` is rewritten to SME (by `TransferWriteToArmSMEZeroLowering`), the type conversion kicks in and attempts to convert `vector<[16]x[16]xi8>`. <--- **What should happen here?** 

IIUC, this type conversion is "thrown away" when `transfer_write` is rewritten, so has no effect on the final result of compilation. Still, it shouldn't be happening at all 🤔 . The output LLVM type is incorrect (`!llvm.array<16 x vector<[16]xi8>>`), but that's a separate issue :) 

My patch for ref:  https://reviews.llvm.org/D154302

**Comment by banach-space** [comment]
I've done a bit more digging and here's a summary.

**NOTE**: The purpose of this discussion is to:
* see whether the `assert` from https://reviews.llvm.org/D154302 makes sense (I believe it does),
*  if yes, find a way to land it.

Also, I am using this one specific `assert` to help with the discussion and make it more specific. But in general, IMHO, it's just a manifestation of larger issue.

## Reproducer 
You will need to apply https://reviews.llvm.org/D154302 to hit the assert that I am trying to introduce and that's meant to prevent the lowering of e.g. `vector<[16]x[16]xi8>` to LLVM.  

```mlir
func.func @transfer_write_2d_zero_i8(%arg0 : memref<?x?xi8>) {
  %c0 = arith.constant 0 : index
  %cst = arith.constant dense<0> : vector<[16]x[16]xi8>
  vector.transfer_write %cst, %arg0[%c0, %c0] {in_bounds = [true, true]} : vector<[16]x[16]xi8>, memref<?x?xi8>
  return
}
```

To trigger the assert:
```
$ mlir-opt repro.mlir -debug-only=dialect-conversion -enable-arm-streaming="mode=locally enable-za"    -convert-vector-to-llvm="enable-arm-sme"
Assertion failed: ((!type.isScalable() || (type.getShape().size() == 1)) && "expected 1-D scalable vector (n-D scalable vectors are not supported)"), function convertVectorType, file TypeConverter.cpp, line 474.
## FULL STACK TRACE HERE
```

## LLVMTypeConversion

IMO, as long as LLVM does not support arrays of scalable vectors, the `assert` that I proposed in https://reviews.llvm.org/D154302 makes a lot of sense. It would be very helpful in capturing unsupported cases before proceeding any further with lowering (very good for debugging). It would be good to find a way to land it.

I will try to quickly explain how it's being triggered in the example above:
1. [TransferWriteToVectorStoreLowering](https://github.com/llvm/llvm-project/blob/2712b2805b47f10b3864ab19a4016ea306126ad7/mlir/lib/Dialect/Vector/Transforms/LowerVectorTransfer.cpp#L531) attempts to rewrite `vector.transfer_write` as `vector.store`,
2. A handful of "legalizers" are run for `vector.store`. Eventually, [VectorLoadStoreConversion](https://github.com/llvm/llvm-project/blob/2712b2805b47f10b3864ab19a4016ea306126ad7/mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp#L231) crashes while running `LLVMTypeConversion`. The crash happens because the operand conversion is attempted before the Op conversion (which would fail anyway). See below. 

Code extracted from [DialectConversion.cpp](
https://github.com/llvm/llvm-project/blob/048764f23a380fd6f8cc562a0008dcc6095fb594/mlir/lib/Transforms/Utils/DialectConversion.cpp#L1648-L1665
):
```cpp
LogicalResult
ConversionPattern::matchAndRewrite(Operation *op,
                                   PatternRewriter &rewriter) const {
  auto &dialectRewriter = static_cast<ConversionPatternRewriter &>(rewriter);
  auto &rewriterImpl = dialectRewriter.getImpl();

  // Track the current conversion pattern type converter in the rewriter.
  llvm::SaveAndRestore currentConverterGuard(rewriterImpl.currentTypeConverter,
                                             getTypeConverter());

  // Remap the operands of the operation.
  SmallVector<Value, 4> operands;
  if (failed(rewriterImpl.remapValues("operand", op->getLoc(), rewriter,
                                      op->getOperands(), operands))) {
    return failure();
  }
  return matchAndRewrite(op, operands, dialectRewriter);
}
```

Note that operands are converted **before** the operation is rewritten. In the reproducer above, the Op conversion for `vector.store` fails even if the type conversion (i.e. operand conversion) succeeds. And when the type conversion does run,  it leads to invalid conversion:

* `vector<[16]x[16]xi8>` --> `!llvm.array<16 x vector<[16]xi8>>`

The assert would help us prevent such code ever being generated.

## Vector to SME conversion

The Vector to SME conversion from https://reviews.llvm.org/D152508 introduces yet another lowering path for:
 * `vector.transfer_write` --> LLVM. 
 
 ATM, that's run after the [Vector-to-Vector conversions](https://github.com/llvm/llvm-project/blob/2712b2805b47f10b3864ab19a4016ea306126ad7/mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVMPass.cpp#L68-L80) mentioned above. 
 
If we reject the idea of an `assert` that I am proposing here (i.e. https://reviews.llvm.org/D154302), then the following happens:
 
1. at some point during the lowering, the following invalid conversion takes place: `vector<[16]x[16]xi8>` --> `!llvm.array<16 x vector<[16]xi8>>`,
2. the above is discarded anyway as the overall conversion of `vector.store` (discussed in the previous section) fails,
3. later, Vector-to-SME lowering takes care of lowering of `vector.transfer_write` onto valid SME code,
4. compilation succeeds.

One could argue that the invalid type conversion "doesn't matter anyway" (since the compilation succeeds), but this is just one very specific case in which "things just work" ™️ . There's no guarantee that it would also "just work" for targets other than SME.

## Possible solution

I tried modifying the order of Vector-to-LLVM conversions so that SME "conversions" [1] are run "early" (in [LowerVectorToLLVMPass::runOnOperation()](https://github.com/llvm/llvm-project/blob/2712b2805b47f10b3864ab19a4016ea306126ad7/mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVMPass.cpp#L65)), but I didn't get far with that. I didn't spend too much time debugging as to me that's a very intrusive change, which I'd like to avoid. We shouldn't require any particular ordering to avoid breakage.

Other, more generic way to resolve this would be to:
*  introduce a custom SME Op to lower `vector.transfer_write` to (`sme.store_tile` ?), and 
* split the lowering of SME custom Ops from the regular "Vector to LLVM" lowering path. 

Although this would be a bit of a shift from the current approach,  I don't see any better option. And the refactoring should be fairly straightforward.

-Andrzej

[1] See [arm_sme::populateVectorTransferLoweringPatterns](https://github.com/llvm/llvm-project/blob/2712b2805b47f10b3864ab19a4016ea306126ad7/mlir/lib/Dialect/ArmSME/Transforms/LowerVectorOps.cpp#L108-L111).

**Comment by c-rhodes** [comment]
Thanks for digging into this Andrzej. I agree the assert makes sense, as it stands MLIR may generate illegal LLVM IR and this assert will catch this early rather than later during translation. It sounds like the simplest option are custom SME ops.

**Comment by dcaballe** [comment]
I agree with the general direction. We decided to skip SME ops in favor of a 1:1 mapping between a Vector op and an LLVM intrinsic. However, the lowering for this `vector.transfer_write` involves creating the zero intrinsic and also a loop that iteratively writes into memory. That's really far from the 1:1 mapping that we anticipated. We should properly model this within the SME dialect.
